library/std/src/sys/platform_version/darwin/public_extern.rs

```diff
@@ -0,0 +1,157 @@
+//! because their implementation is quite complex, using allocation, environment variables, file
+//! access and dynamic library loading (and emitting all of this into every binary).
+//!
+//! The reason why Apple chose to not do that originally is lost to the sands of time, but a good
+//! reason would be that implementing it as part of `compiler-rt` allowed them to back-deploy this
+//! to older OSes immediately.
+//!
+//! In Rust's case, while we may provide a feature similar to `@available` in the future, we will
+//! probably do so as a macro exposed by `std` (and not as a compiler builtin). So implementing this
+//! in `std` makes sense, since then we can implement it using `std` utilities, and we can avoid
+//! having `compiler-builtins` depend on `libSystem.dylib`.
+//!
+//! This does mean that users that attempt to link C/Objective-C/Swift code _and_ use `#![no_std]`
+//! in all their crates may get a linker error because these symbols are missing. Using `no_std` is
+//! quite uncommon on Apple systems though, so it's probably fine to not support this use-case.
+//!
+//! The workaround would be to link `libclang_rt.osx.a` or otherwise use Clang's `compiler-rt`.
+//!
+//! See also discussion in <https://github.com/rust-lang/compiler-builtins/pull/794>.
+//!
+//! ## Implementation details
+//!
+//! NOTE: Since macOS 10.15, `libSystem.dylib` _has_ actually provided the undocumented
+//! `_availability_version_check` via `libxpc` for doing the version lookup (zippered, which is why
+//! it requires a platform parameter to differentiate between macOS and Mac Catalyst), though its
+//! usage may be a bit dangerous, see:
+//! - https://reviews.llvm.org/D150397
+//! - https://github.com/llvm/llvm-project/issues/64227
+//!
+//! Besides, we'd need to implement the version lookup via. PList to support older versions anyhow,
+//! so we might as well use that everywhere (since it can also be optimized more after inlining).
+
+#![allow(non_snake_case)]
+
+use super::{OSVersion, current_version, pack_os_version};
+
+/// Whether the current platform's OS version is higher than or equal to the given version.
+///
+/// The first argument is the _base_ Mach-O platform (i.e. `PLATFORM_MACOS`, `PLATFORM_IOS`, etc.,
+/// but not `PLATFORM_IOSSIMULATOR` or `PLATFORM_MACCATALYST`) of the invoking binary.
+///
+/// Arguments are specified statically by Clang. Inlining with LTO should allow the versions to be
+/// combined into a single `u32`, which should make comparisons faster, and should make the
+/// `BASE_TARGET_PLATFORM` check a no-op.
+//
+// SAFETY: The signature is the same as what Clang expects, and we export weakly to allow linking
+// both this and `libclang_rt.*.a`, similar to how `compiler-builtins` does it:
+// https://github.com/rust-lang/compiler-builtins/blob/0.1.113/src/macros.rs#L494
+#[cfg_attr(not(feature = "compiler-builtins-mangled-names"), unsafe(no_mangle), linkage = "weak")]
```

Comment:
This would also cause these symbols to be exported from cdylibs which should not export any non-user exported symbols.

Reply:
Also `feature = "compiler-builtins-mangled-names"` doesn't exist inside libstd.

Reply:
> This would also cause these symbols to be exported from cdylibs which should not export any non-user exported symbols.

I'm curious what you suggest I do instead / how I make this do the same thing as it'd do if it were part of `compiler-builtins`?

> Also `feature = "compiler-builtins-mangled-names"` doesn't exist inside libstd.

It does:
https://github.com/rust-lang/rust/blob/7d6f41b08e730572b196d6c588364973bb3cb12b/library/std/Cargo.toml#L101

But might not be used in the same way?

Reply:
> I'm curious what you suggest I do instead / how I make this do the same thing as it'd do if it were part of compiler-builtins?

`#[rustc_std_internal_symbol]` + a special case to disable mangling for these symbols should work. See also how rust_eh_personality has a special case for `#[rustc_std_internal_symbol]`.

Reply:
I've done this now, see the last commit, thanks for the guidance!